text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,484 @@
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
+might pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the "Create" span of the messages forwarded, thus those
+spans can link to zero, one, or multiple "Create" spans.
+
+#### Settlement
+
+Messages can be settled in a variety of different ways. In some cases, messages
+are not settled at all (fire-and-forget), in other cases settlement operations
+are triggered manually by the user, and in callback scenarios settlement can be
+automatically triggered by messaging SDKs based on return values of callbacks.
+
+A "Settle" span should be created for every settlement operation, no matter if
+the settlement operation was manually triggered by the user or automatically
+triggered by SDKs. SDKs will in some cases auto-settle messages in
+push-scenarios when messages are delivered via callbacks. In this case, it is
+recommended to create a parent span, so that a "Settle" span will be a sibling
+of the related "Deliver" span.
+
+Alternatively, an event can be created instead of a "Settle" span. Events could
+be added to "Deliver" spans or to ambient spans.
+
+"Settle" spans may link to "Create" spans of the messages that are settled,
+however, for some settlement scenarios this is not feasible or possible.
+
+## Proposed changes and additions to the messaging semantic conventions
+
+### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One or more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
```

Comment:
can you elaborate why it's `INTERNAL` if "Create" spans are present please?

Reply:
If both "publish" and "create" spans are present, and both would be of kind `PRODUCER`, then one would end up with nested producer spans. This is something we want to avoid, as it creates ambiguity and makes it harder to analyze traces.

In addition to avoid nested span, we can also make this statement:

> Setting span kinds according to this table ensures, that span links between consumers and producers always go from a `PRODUCER` span on the producer side to a `CONSUMER` span on the consumer side. This allows analysis tools to interpret linked traces without the need for additional semantic hints.

Having a single producer span for each message, and having a single consumer span for each time a message is received from an intermediary makes it much easier to interpret traces.